File: PANPRSnext/R/PANPRS.R
#' Run the package on the provided data set
#' @export
test_pkg <- function(
  debug_output = FALSE
) {
  data("summaryZ")
  data("Nvec")
  data("plinkLD")
  data("funcIndex")
  output <- gsfPEN_R(
    summary_z = summaryZ, # nolint: object_usage_linter.
    n_vec = Nvec, # nolint: object_usage_linter.
    plinkLD = plinkLD, # nolint: object_usage_linter.
    func_index = funcIndex, # nolint: object_usage_linter.
    debug_output = debug_output
  )

  return(output)
}

#' Run the package on the provided data set
#' DEBUG VERSION
#' @export
debug_test_pkg <- function(
  summaryZ,
  funcIndex,
  debug_output = FALSE,
  sub_tuning = 1,
  lambda_vec_limit_len = c(1.5, 1)
) {
  data("Nvec")
  data("plinkLD")
  output <- gsfPEN_R(
    summary_z = summaryZ, # nolint: object_usage_linter.
    n_vec = Nvec, # nolint: object_usage_linter.
    plinkLD = plinkLD, # nolint: object_usage_linter.
    func_index = funcIndex, # nolint: object_usage_linter.
    sub_tuning = sub_tuning,
    lambda_vec_limit_len = lambda_vec_limit_len,
    debug_output = debug_output
  )

  return(output)
}

#' Main Function
#' @export
gsfPEN_R <- function(
    summary_z,
    n_vec,
    plinkLD,
    func_index,
    n_iter = 1000,
    upper_val = NULL,
    breaking = 1,
    z_scale = 1,
    tuning_matrix = NULL,
    p_threshold = NULL,
    p_threshold_params = c(0.5, 10^-4, 4),
    tau_factor = c(1 / 25, 1, 3),
    len_lim_lambda = 4,
    sub_tuning = 4,
    lim_lambda = c(0.5, 0.9),
    len_lambda = 4,
    lambda_vec = NULL,
    lambda_vec_limit_len = c(1.5, 3),
    df_max = NULL,
    debug_output = FALSE
) {

  time <- proc.time()

  if (z_scale != 1) {
    error("Tuning values set-up for multiple traits analysis requires z_scale=1.") # nolint: object_usage_linter.
  }

  num_func <- ncol(func_index)

  P <- nrow(summary_z)
  Q <- length(n_vec)

  summary_betas <- matrix(0, nrow = P, ncol = Q)
  SD_vec <- matrix(0, nrow = P, ncol = Q)

  for (ii in 1:Q) {
    summary_betas[, ii] <- summary_z[, ii] / sqrt(n_vec[ii])
    SD_vec[, ii] <- 1 / sqrt(n_vec[ii])
  }

  rownames(summary_betas) <- rownames(summary_z)

  if (is.null(df_max)) {
    df_max <- ceiling(0.7 * P)
  }

  if (is.null(p_threshold)) {
    p_threshold <- seq(p_threshold_params[1], p_threshold_params[2], length.out = p_threshold_params[3])
  }

  if (any(c(is.null(tuning_matrix), is.null(lambda_vec)))) {
    median_val <- median(apply(abs(summary_betas), 1, sum), na.rm = TRUE)
    tau_vec <- sort(median_val * tau_factor)
    lim_lambda <- quantile(abs(summary_z[, 1]), lim_lambda)

    output <- Tuning_setup_group_func(
      lambda_vec,
      lambda_vec_limit_len,
      p_threshold,
      num_func,
      tau_vec,
      sub_tuning,
      lim_lambda,
      len_lambda,
      len_lim_lambda,
      median_val
    )

    func_lambda <- output$func_lambda
    lambda_vec <- output$lambda_vec
    tuning_matrix <- output$tuning_matrix

    rm(output)
  } else {
    func_lambda <- permutations( # nolint: object_usage_linter.
      length(lambda_vec),
      num_func,
      repeats.allowed = TRUE
    ) - 1
  }

  beta_index <- c(seq_len(nrow(summary_betas))) - 1
  SNP_names <- rownames(summary_betas)

  # Takes only the SNPs that are present in both the PLINK data set and the GWAs
  ld_J <- PlinkLD_transform(plinkLD, SNP_names)

  J_id_matrix <- matrix(nrow = nrow(ld_J), ncol = 2)

  mat1 <- match(ld_J[, 1], SNP_names)
  mat2 <- match(ld_J[, 2], SNP_names)

  J_id_matrix[, 1] <- beta_index[mat1]
  J_id_matrix[, 2] <- beta_index[mat2]

  ld_J[, 1] <- J_id_matrix[, 1]
  ld_J[, 2] <- J_id_matrix[, 2]

  od <- order(J_id_matrix[, 1], J_id_matrix[, 2], decreasing = FALSE)
  ld_J <- ld_J[od, ]


  wind <- which(!beta_index %in% ld_J[, 1])
  num_indices <- length(wind)

  index_J <- -1
  if (num_indices > 0) {
    index_J <- beta_index[wind]
  }

  counts <- table(ld_J[, 1])
  num_SNP <- length(counts)

  index_S <- c(0, cumsum(counts)[-num_SNP])
  index_E <- cumsum(counts) - 1

  index_matrix <- matrix(nrow = num_SNP, ncol = 3)
  index_matrix[, 1] <- as.numeric(names(counts))

  index_matrix[, 2] <- index_S
  index_matrix[, 3] <- index_E

  ld_vec <- ld_J[, 3]
  ld_J <- ld_J[, 2]


  if (is.null(upper_val)) {
    upper_val <- ceiling(max(abs(summary_betas), na.rm = TRUE) * 50)
  }

  if (nrow(func_index) != nrow(summary_betas)) {
    stop("nrow of summary_betas and row of func_index do not match.")
  }


  z_matrix <- matrix(1 - unlist(func_index), nrow = nrow(func_index), ncol = ncol(func_index), byrow = TRUE)
  rownames(z_matrix) <- rownames(func_index)

  sum_func_index <- apply(z_matrix, 1, sum)
  Ifunc_SNP <- rep(0, P)
  Ifunc_SNP[which(sum_func_index != 0)] <- 1

  lambda0_vec <- abs(-qnorm(p_threshold / 2))


  nrow_index_matrix <- nrow(index_matrix)
  ncol_index_matrix <- ncol(index_matrix)

  nrow_z_matrix <- nrow(z_matrix)
  ncol_z_matrix <- ncol(z_matrix)

  nrow_func_lambda <- nrow(func_lambda)
  ncol_func_lambda <- ncol(func_lambda)

  nrow_tuning_matrix <- nrow(tuning_matrix)
  ncol_tuning_matrix <- ncol(tuning_matrix)

  nrow_all_tuning_matrix <- nrow_tuning_matrix * length(p_threshold) * nrow_func_lambda
  ncol_all_tuning_matrix <- (num_func + 1) + 2

  nrow_beta_matrix <- nrow_all_tuning_matrix
  ncol_beta_matrix <- P * Q

  dims <- c(
    num_SNP,                # 1
    P,                      # 2
    Q,                      # 3
    nrow_index_matrix,      # 4
    ncol_index_matrix,      # 5
    nrow_z_matrix,          # 6
    ncol_z_matrix,          # 7
    nrow_func_lambda,       # 8
    ncol_func_lambda,       # 9
    nrow_tuning_matrix,     # 10
    ncol_tuning_matrix,     # 11
    nrow_all_tuning_matrix, # 12 == num_tuning
    ncol_all_tuning_matrix, # 13
    nrow_beta_matrix,       # 14
    ncol_beta_matrix        # 15
  )

  params <- c(
    upper_val,              # 1
    n_iter,                 # 2
    breaking,               # 3
    z_scale,                # 4
    df_max,                 # 5
    length(p_threshold),    # 6
    num_indices             # 7
  )

  print(paste0("Number of total tuning combinations = ", nrow_all_tuning_matrix))

  Z <- gsfPEN_cpp(
    summary_betas,
    ld_J,
    index_matrix,
    index_J,
    ld_vec,
    SD_vec,
    tuning_matrix,
    lambda0_vec,
    z_matrix,
    lambda_vec,
    func_lambda,
    Ifunc_SNP,
    dims,
    params
  )

  beta_matrix <- Z$beta_matrix
  colnames(beta_matrix) <- paste0(
    rep(SNP_names, times = Q),
    ".trait",
    rep(c(1:Q), each = P)
  )

  all_tuning_matrix <- Z$all_tuning_matrix
  colnames(all_tuning_matrix) <- c(
    "lambda0",
    paste0("lambdaf", c(1:num_func)),
    "lamda2",
    "tau2"
  )

  num_iter_vec <- Z$num_iter_vec

  # Remove the tuning combinations that did not converge (correspons to -2 in num_iter_vec)
  if (!debug_output) {
    print("Removing tuning combinations that did not converge.")
    converge_index <- which(num_iter_vec > 0)
    num_iter_vec <- num_iter_vec[converge_index]
    beta_matrix <- beta_matrix[converge_index, ]
    all_tuning_matrix <- all_tuning_matrix[converge_index, ]
  }

  output <- Clean_results(
    beta_matrix = beta_matrix,
    num_iter_vec = num_iter_vec,
    all_tuning_matrix = all_tuning_matrix
  )

  time <- proc.time() - time
  print(paste0("Time elapsed: ", time[3], " seconds"))

  return(output)
}

File: PANPRSnext/src/PANPRS.cpp
#include <RcppArmadillo.h>

#include <stdio.h>
#include <stdlib.h>

#include "PANPRS.hpp"

Rcpp::List gsfPEN_cpp(
    arma::Mat<double> summary_betas,
    arma::Col<int> ld_J,
    arma::Mat<int> index_matrix,
    arma::Col<int> index_J,
    arma::Col<double> ld_vec,
    arma::Mat<double> SD_vec,
    arma::Mat<double> tuning_matrix,
    arma::Col<double> lambda0_vec,
    arma::Mat<double> z_matrix,
    arma::Col<double> lambda_vec,
    arma::Mat<int> func_lambda,
    arma::Col<int> Ifunc_SNP,
    arma::Col<int> dims,
    arma::Col<double> params)
{
  int num_SNP = dims(0);

  int P = dims(1);
  int Q = dims(2);

  int nrow_index_matrix = dims(3);
  int ncol_index_matrix = dims(4);

  int nrow_z_matrix = dims(5);
  int ncol_z_matrix = dims(6);

  int nrow_func_lambda = dims(7);
  int ncol_func_lambda = dims(8);

  int nrow_tuning_matrix = dims(9);
  int ncol_tuning_matrix = dims(10);

  int nrow_all_tuning_matrix = dims(11);
  int ncol_all_tuning_matrix = dims(12);

  int nrow_beta_matrix = dims(13);
  int ncol_beta_matrix = dims(14);

  double upper_val = params(0);
  int num_iter = params(1);
  int breaking = params(2);
  int z_scale = params(3);
  int df_max = params(4);
  int leng_p_threshold = params(5);
  int num_indices = params(6);

  double epsilon = 0.0001;

  arma::Col<int> num_iter_vec(nrow_all_tuning_matrix, arma::fill::zeros);
  arma::Col<double> temp_lambda_vec(ncol_func_lambda, arma::fill::zeros);
  arma::Col<double> sum_betas(P, arma::fill::zeros);

  arma::Mat<double> all_tuning_matrix(nrow_all_tuning_matrix, ncol_all_tuning_matrix, arma::fill::zeros);
  // arma::Mat<double> beta_matrix(nrow_beta_matrix, ncol_beta_matrix, arma::fill::zeros);
  arma::SpMat<double> beta_matrix(nrow_beta_matrix, ncol_beta_matrix);
  arma::Mat<double> joint_b_matrix(P, Q, arma::fill::zeros);
  arma::Mat<double> temp_b_matrix(P, Q, arma::fill::zeros);
  arma::Mat<int> skip(P, Q, arma::fill::zeros);

  GetRNGstate();

  // len_p_Thresold = 4
  // By default, p.Threshold = seq(0.5, 10^-4, length.out=4)
  int max_tuning_index = nrow_all_tuning_matrix * nrow_func_lambda * nrow_tuning_matrix;
  int tuning_index = -1;
  for (int threshold_index = 0; threshold_index < leng_p_threshold; threshold_index++)
  {
    for (int tun_idx_1 = 0; tun_idx_1 < nrow_func_lambda; tun_idx_1++)
    {
      for (int tun_idx_2 = 0; tun_idx_2 < nrow_tuning_matrix; tun_idx_2++)
      {
        tuning_index++;

        for (int func_index = 0; func_index < ncol_func_lambda; func_index++)
        {
          temp_lambda_vec(func_index) = lambda_vec(func_lambda(tun_idx_1, func_index));
          all_tuning_matrix(tuning_index, func_index + 1) = temp_lambda_vec(func_index);
        }

        // lambda0vec = abs(-qnorm(p.Threshold/2)) -> 0.67 0.96 1.38 3.89
        double lambda0 = lambda0_vec(threshold_index);
        all_tuning_matrix(tuning_index, 0) = lambda0;

        double lambda2 = tuning_matrix(tun_idx_2, 2);
        all_tuning_matrix(tuning_index, ncol_all_tuning_matrix - 2) = lambda2;

        double tau2 = tuning_matrix(tun_idx_2, 3);
        all_tuning_matrix(tuning_index, ncol_all_tuning_matrix - 1) = tau2;

        bool converges = true;
        for (int n = 1; n <= num_iter; n++)
        {
          // This is non zero when there are Z-scores (summary_betas) that are not in the linkage data (LdJ)
          // Realistically, this is always 0 given the R interfaces only selects those that are present in both
          if (num_indices != 0)
          {
            // num_indices = 3, 1: SNP_A, 2: SNP_B, 3: R
            for (int i = 0; i < num_indices; i++)
            {
              int j = index_J(i);
              double lambda1 = lambda0;

              if (Ifunc_SNP(j) == 1)
              {
                for (int func_index = 0; func_index < ncol_func_lambda; func_index++)
                {
                  lambda1 = lambda1 + z_matrix(j, func_index) * temp_lambda_vec(func_index);
                }
              }

              for (int q = 0; q < Q; q++)
              {
                double bj_bar = summary_betas(j, q);
                if (bj_bar != 0.0)
                {
                  double threshold;
                  if (Q == 1)
                    threshold = lambda1;
                  else
                    threshold = lambda1 + lambda2 / (sum_betas(j) + tau2);

                  // SDvec is 1/sqrt(sample size of curr. GWA)
                  if (z_scale == 1)
                    threshold *= SD_vec(j, q);

                  if (bj_bar > threshold)
                    joint_b_matrix(j, q) = bj_bar - threshold;
                  else if (bj_bar < -threshold)
                    joint_b_matrix(j, q) = bj_bar + threshold;
                  else
                    joint_b_matrix(j, q) = 0.0;
                }

                if (summary_betas(j, q) * joint_b_matrix(j, q) < 0)
                {
                  Rprintf("summary_betas[j]=%d\n", j);
                  Rprintf("summary_betas[q]=%d\n", q);
                  Rprintf("summary_betas[j][q]=%e\n", summary_betas(j, q));
                  Rprintf("joint_b_matrix[j][q]=%e\n", joint_b_matrix(j, q));
                  perror("sign inverse");
                }
              }
            }
          }

          // Everything up to this point is only done on the first iteration, of course
          // once for each set of tuning params (standard and functional)
          // Inserting a “#pragma omp parallel for” right here (with some other boilerplate) should be
          // All that is necessary to parallelize it (to start) (will test this)
          // Since there are lots of nested for loops, it may be possible to parallelize it further with
          // “collapse(n)” where n is the number of nested loops to parallelize
          for (int p = 0; p < num_SNP; p++)
          {
            int j = index_matrix(p, 0);
            double lambda1 = lambda0;

            if (Ifunc_SNP(j) == 1)
            {
              for (int func_index = 0; func_index < ncol_func_lambda; func_index++)
              {
                lambda1 += z_matrix(j, func_index) * temp_lambda_vec(func_index);
              }
            }

            for (int q = 0; q < Q; q++)
            {
              if (skip(j, q) == 0)
              {
                if (summary_betas(j, q) != 0.0)
                {
                  double bj_bar = summary_betas(j, q);
                  int i = index_matrix(p, 1);

                  for (int i = index_matrix(p, 1); i < index_matrix(p, 2) + 1; i++)
                  {
                    bj_bar -= ld_vec(i) * joint_b_matrix(ld_J(i), q);
                  }
                  // if (tuning_index == 18 && n == 1) {
                  //   printf("q: %d, i: %d, ld_vec(i): %f\n", q, i, ld_vec(i), joint_b_matrix(ld_J(i), q));
                  //   printf("ld_J(i): %d, joint_b_matrix(ld_J(i), q): %f\n", ld_J(i), joint_b_matrix(ld_J(i), q));
                  //   printf("bj_bar: %f\n", bj_bar);
                  // }

                  // if (tuning_index == 18 && n == 1) printf("lambda1: %f\n", lambda1);
                  double threshold;
                  if (Q == 1)
                    threshold = lambda1;
                  else
                    threshold = lambda1 + lambda2 / (sum_betas(j) + tau2);

                  // if (tuning_index == 18 && n == 1) {
                  //   printf("threshold: %f\n", threshold);
                  //   printf("sum_betas(j): %f\n", sum_betas(j));
                  // }

                  if (fabs(bj_bar) > upper_val)
                  {
                    if (breaking == 1)
                    {
                      num_iter_vec(tuning_index) = -1;
                      break;
                    }
                    else
                    {
                      bj_bar = 0.0;
                      skip(j, q) = 1;
                    }
                  }

                  if (z_scale == 1)
                    threshold *= SD_vec(j, q);

                  if (bj_bar > threshold)
                    joint_b_matrix(j, q) = bj_bar - threshold;
                  else if (bj_bar < -threshold)
                    joint_b_matrix(j, q) = bj_bar + threshold;
                  else
                    joint_b_matrix(j, q) = 0.0;
                }
                else
                {
                  joint_b_matrix(j, q) = 0.0;
                }
              }
              else
              {
                joint_b_matrix(j, q) = 0.0;
              }
            }
          }

          // Actual iteration that checks convergence
          bool found = false;
          int df_q = 0;
          for (int p = 0; p < P; p++)
          {
            for (int q = 0; q < Q; q++)
            {
              if (fabs(joint_b_matrix(p, q)) > upper_val)
                skip(p, q) = 1;

              if (q == 0)
              {
                if (fabs(joint_b_matrix(p, q)) != 0)
                  df_q++;

                if (df_q > df_max)
                {
                  converges = false;
                  break;
                }
              }

              if (fabs(temp_b_matrix(p, q) - joint_b_matrix(p, q)) > epsilon)
              {
                found = true;
                break;
              }
            }
          }

          if (!converges)
          {
            num_iter_vec(tuning_index) = -2;
            break;
          }

          if (!found)
          {
            int beta_idx = 0;
            for (int q = 0; q < Q; q++)
            {
              for (int p = 0; p < P; p++)
              {
                if (joint_b_matrix(p, q) != 0.0)
                  beta_matrix(tuning_index, beta_idx) = joint_b_matrix(p, q);
                beta_idx++;
              }
            }

            num_iter_vec(tuning_index) = n;
            break;
          }

          for (int p = 0; p < num_SNP; p++)
          {
            int j = index_matrix(p, 0);
            sum_betas(j) = 0.0;

            for (int q = 0; q < Q; q++)
            {
              temp_b_matrix(j, q) = joint_b_matrix(j, q);
              sum_betas(j) += fabs(joint_b_matrix(j, q));
            }
          }

          if (num_indices != 0)
          {
            for (int i = 0; i < num_indices; i++)
            {
              int j = index_J(i);
              sum_betas(j) = 0.0;
              for (int q = 0; q < Q; q++)
              {
                temp_b_matrix(j, q) = joint_b_matrix(j, q);
                sum_betas(j) += fabs(joint_b_matrix(j, q));
              }
            }
          }

          num_iter_vec(tuning_index) = n;
        }
      }
    }
  }
  PutRNGstate();

  return Rcpp::List::create(
      Rcpp::Named("beta_matrix") = beta_matrix,
      Rcpp::Named("num_iter_vec") = num_iter_vec,
      Rcpp::Named("all_tuning_matrix") = all_tuning_matrix);
}
